### ISSUE TITLE
Implementation of PatchHDCELoss

### ISSUE THREAD

**Issue by wcshin-git** [jcy132/Hneg_SRC#3]
Hi, thanks for your great work!

I have a question about the PatchHDCELoss implementation in the Single-modal setting.

In models/hDCE.py,
```python
class PatchHDCELoss(nn.Module):
   ...
   ...
        diagonal = torch.eye(npatches, device=feat_q.device, dtype=self.mask_dtype)[None, :, :]
        l_neg_curbatch.masked_fill_(diagonal, -10.0) 
        l_neg = l_neg_curbatch.view(-1, npatches) # [npatches, npatches]

        logits = (l_neg-l_pos)/self.opt.nce_T # [npatches, npatches]
        v = torch.logsumexp(logits, dim=1)    # [npatches]
        loss_vec = torch.exp(v-v.detach())    # [npatches] all elements are 1.0

        # for monitoring
        out_dummy = torch.cat((l_pos, l_neg), dim=1) / self.opt.nce_T # [npatches, npatches+1]
        CELoss_dummy = self.cross_entropy_loss(out_dummy, torch.zeros(out_dummy.size(0), dtype=torch.long, device=feat_q.device))

        loss = loss_vec.mean()-1+CELoss_dummy.detach()

        return loss
```
I can't understand your intentions in this implementation. 
Which equation in your paper did you implement? 
I cannot match this loss to a equation in the paper.
I think I missed something. Please give me some explanation.
Thank you!

**Comment by jcy132** [comment]
It is from Decoupled infoNCE(DCE) loss, which has same form of FlatCLR. Please refer the Code in  https://github.com/Junya-Chen/FlatCLR

**Comment by Healingl** [comment]
> ```python
> class PatchHDCELoss(nn.Module):
>    ...
>    ...
>         diagonal = torch.eye(npatches, device=feat_q.device, dtype=self.mask_dtype)[None, :, :]
>         l_neg_curbatch.masked_fill_(diagonal, -10.0) 
>         l_neg = l_neg_curbatch.view(-1, npatches) # [npatches, npatches]
> 
>         logits = (l_neg-l_pos)/self.opt.nce_T # [npatches, npatches]
>         v = torch.logsumexp(logits, dim=1)    # [npatches]
>         loss_vec = torch.exp(v-v.detach())    # [npatches] all elements are 1.0
> 
>         # for monitoring
>         out_dummy = torch.cat((l_pos, l_neg), dim=1) / self.opt.nce_T # [npatches, npatches+1]
>         CELoss_dummy = self.cross_entropy_loss(out_dummy, torch.zeros(out_dummy.size(0), dtype=torch.long, device=feat_q.device))
> 
>         loss = loss_vec.mean()-1+CELoss_dummy.detach()
> 
>         return loss
> ```

Hi, I don't understand 'loss = loss_vec.mean()-1+CELoss_dummy.detach()', which seems to output the same results of 'CELoss_dummy.detach()'. Because 'loss_vec = torch.exp(v-v.detach())' is a tensor with all values of 1, I think  ‘loss_vec.mean()-1’ is a tensor with all values of 0. I don't know if this loss is same with infoNCE?
